# Introduction to RNA-seq
## Teaching Materials
This repository contains teaching materials from hbctraining but has been modified to support students enrolled in MMG232: Advanced Bioinformatics at the University of Vermont. The full citation for this work is provided below.
## Description
This repository has materials for an Introduction to single cell RNA-sequencing data analysis. This unit focuses on on how to design a single-cell RNA-seq experiment, and how to efficiently manage and analyze the data starting from count matrices. The steps to generate count matrices from raw data will be discussed, however, the steps to generate count matrices from raw data (FASTQ files) will not be performed. In this unit, we will focus on using the Seurat package using R/RStudio on VACC Open OneDemand. Working knowledge of R/Rstudio is required.
## Learning Objectives
1. Explain common considerations when designing a single-cell RNA-seq experiment
2. Discuss the steps involved in taking raw single-cell RNA-sequencing data and generating a count (gene expression) matrix
3. Compute and assess QC metrics at every step in the workflow
4. Cluster cells based on expression data and derive the identity of the different cell types present
5. Identify the expression pattern of specific gene(s) at the single cell level
## To begin: Click on Class schedule
