# Introduction to single-cell RNA-seq

## Teaching Materials

This repository contains teaching materials from hbctraining but has been modified to support students enrolled in MMG232: Advanced Bioinformatics at the University of Vermont. The full citation for this work is provided below.

## Description

This repository has materials for an Introduction to single-cell RNA-sequencing data analysis. This unit focuses on designing a single-cell RNA-seq experiment and managing and analyzing the data starting from count matrices efficiently. The steps to generate count matrices from raw data will be discussed; however, the steps to create count matrices from raw data (FASTQ files) will not be performed. In this unit, we will use the Seurat package using R/RStudio on VACC Open OnDemand. Working knowledge of R/Rstudio is required.

## Learning Objectives

1. Explain common considerations when designing a single-cell RNA-seq experiment
2. Discuss the steps involved in taking raw single-cell RNA-sequencing data and generating a count (gene expression) matrix
3. Compute and assess QC metrics at every step in the workflow
4. Cluster cells based on expression data and derive the identity of the different cell types present
5. Identify the expression pattern of specific gene(s) at the single cell level

## To begin: Click on Class schedule 

[Class Schedule](/Schedule/README.md)

## Class Lectures

[Day 1: Introduction to single-cell RNA sequencing](/Lessons/1_intro_to_scRNA-seq.md)

[Day 2: Data loading and QC filtering of scRNA-Seq data](/Lessons/2_data_loading_and_QC_filtering.md)
      
[Day 3: Data integration and clustering](/Lessons/3_integration_and_clustering.md)

[Day 4: Data integration and clustering](/Lessons/4_data_visualization_interpretation.md)

                     
### Citation

To cite material from this course in your publications, please use:

> Jihe Liu, William J. Gammerdinger, Meeta Mistry, Mary E. Piper, & Radhika S. Khetani. (2022, January 6). hbctraining/Intro-to-shell-flipped: Shell and HPC Lessons from HCBC (first release). Zenodo. https://doi.org/10.5281/zenodo.5826091

A lot of time and effort went into the preparation of these materials. Citations help us understand the needs of the community, gain recognition for our work, and attract further funding to support our teaching activities. Thank you for citing this material if it helped you in your data analysis.

---
These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.

Some materials used in these lessons were derived from work that is Copyright Â© Data Carpentry (http://datacarpentry.org/). 
All Data Carpentry instructional material is made available under the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0).
