---
title: "R Notebook - 1 Data input and QC filtering and preliminary_clustering"
output: html_notebook
---
#https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
#https://satijalab.org/seurat/articles/merge_vignette.html
#https://satijalab.org/seurat/articles/integration_introduction.html

##### Install required Packages listed below from Bioconductor (No need to run if the package is already installed) 
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("XVector",force = TRUE)
BiocManager::install("multtest")
BiocManager::install("glmGamPo")
```

##### Install required Packages listed below from CRAN (No need to run if the package is already installed) 
```{r}
install.packages('tidyverse')
install.packages('Matrix')
install.packages('RCurl')
install.packages('scales')
install.packages('metap')
install.packages('Seurat')
install.packages("ggplot2")
```

##### Load required libraries
```{r}
library(XVector)
library(Seurat)
library(tidyverse)
library(Matrix)
library(RCurl)
library(scales)
```


##### single-cell RNA-seq count data ("cellranger count" output)
```{r load data}
data1 <- Read10X("10x_human_1_filtered_feature_bc_matrix")
data2 <- Read10X("10x_human_2_filtered_feature_bc_matrix")
data3 <- Read10X("10x_human_3_filtered_feature_bc_matrix")
data4 <- Read10X("10x_human_4_filtered_feature_bc_matrix")
```


##### Create seurat objects
```{r}
data_seurat1 <- CreateSeuratObject(counts = data1, project = "Human_1", min.cells = 3, min.features = 200)
data_seurat2 <- CreateSeuratObject(counts = data2, project = "Human_2", min.cells = 3, min.features = 200)
data_seurat3 <- CreateSeuratObject(counts = data3, project = "Human_3", min.cells = 3, min.features = 200)
data_seurat4 <- CreateSeuratObject(counts = data4, project = "Human_4", min.cells = 3, min.features = 200)
```

##### Merge Seurat objects
```{r}
data_merged <- merge(data_seurat1, y = c(data_seurat2, data_seurat3, data_seurat4), add.cell.ids = c("H1", "H2", "H3","H4"), project = "Human_1234")
```


##### Calculate percent.MT and percent.RIBO
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
#First add column with mitochondiral gene expression
data_merged[["percent.MT"]] <- PercentageFeatureSet(data_merged, pattern = "^MT-")
#Add column with ribosomal gene expression
data_merged[["percent.RIBO"]] <- PercentageFeatureSet(data_merged, pattern = "^RP[SL]")
```


##### Visualize QC metrics
```{r}
VlnPlot(data_merged, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot(data_merged, features = c("percent.MT","percent.RIBO"), ncol = 2)
FeatureScatter(data_merged, feature1 = "percent.RIBO", feature2 = "percent.MT")
FeatureScatter(data_merged, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

##### Save upto this point (optional)
```{r}
saveRDS(data_merged, "data_merged.rds")
```

##### Read the saved data.merged file (optional, this is an alternative starting point)
```{r}
data_merged <- readRDS("data_merged.rds")
```

##### Perform filtering and remake QC plots again on filtered cells
```{r}
#Perform filtering
data_filtered <- subset(data_merged, subset = nFeature_RNA > 700 & nCount_RNA < 100000 & percent.MT < 15 & percent.RIBO > 3)
VlnPlot(data_filtered, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot(data_filtered, features = c("percent.MT","percent.RIBO"), ncol = 2)
FeatureScatter(data_filtered, feature1 = "percent.RIBO", feature2 = "percent.MT")
FeatureScatter(data_filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```

##### Save the filtered seurat object as a RDS file (IMPORTANT!!!)
```{r}
saveRDS(data_filtered, file = "data_filtered.rds")
```

##### Read the saved data.filtered file (optional, this is an alternative starting point)
```{r}
data_filtered <- readRDS("data_filtered.rds")
```

##### Normalizing the data
```{r}
data_norm <- NormalizeData(data_filtered, normalization.method = "LogNormalize", scale.factor = 10000)
```

##### Identification of highly variable features (feature selection)
```{r}
data_norm <- FindVariableFeatures(data_norm, selection.method = "vst", nfeatures = 2000)
```

##### Scaling the data

```{r}
data_norm <- ScaleData(data_norm) #data_norm <- ScaleData(data_norm, vars.to.regress = c("S.Score", "G2M.Score"))
```

##### Cell cycle scoring and visualization (optional)
```{r}
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

data_norm <- CellCycleScoring(data_norm, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)

#Running a PCA on cell cycle genes to evaluate the effect of cell cycle genes
data_norm <- RunPCA(data_norm, features = c(s.genes, g2m.genes))
DimPlot(data_norm)
```

##### Perform linear dimensional reduction and examine the results
```{r}
#Perform linear regression
data_dims<- RunPCA(data_norm, features = VariableFeatures(object = data_norm))

#Rank principle components
ElbowPlot(data_dims)
```

##### Cluster the cells and run Run non-linear dimensional reduction (UMAP)

```{r}
data_clust <- FindNeighbors(data_dims, dims = 1:15)
data_clust <- FindClusters(data_clust, resolution = 0.5)
data_clust <- RunUMAP(data_clust, dims = 1:15)
DimPlot(data_clust, reduction = "umap", label = TRUE)
ggsave("Human_PDA_preliminary_clustering.png",  height=8, width=12, units='in', dpi = 300, bg = "transparent", device='png')
```

##### Check for biological variation/batch effect and cell cycle scoring
```{r}
# Visualize "Phase" on clustered cells
Idents(data_clust) <- "Phase"
DimPlot(data_clust, pt.size = 0.5, label = FALSE, label.size = 5)
ggsave("Human_PDA_preliminary_clustering_cellcyclePhase.png",  height=8, width=12, units='in', dpi = 300, bg = "transparent", device='png')

# Visualize biological variation/batch effect of the samples on clustered cells
DimPlot(data_clust, split.by = "orig.ident", group.by = "seurat_clusters", pt.size = 0.5, label = TRUE, label.size = 5, ncol = 2)
ggsave("Human_PDA_preliminary_clustering_batcheffect.png",  height=8, width=12, units='in', dpi = 300, bg = "transparent", device='png')
```

##### save the preliminary clustered file (this is just a preliminary clustering. In the next class, we will start with the "data_filtered" file)

```{r}
saveRDS(data_clust, "data_preliminary_clustered.rds")
```

------End------